server/gitrest/packages/gitrest-base/src/routes/summaries.ts

```diff
@@ -148,44 +148,46 @@ async function createSummary(
 
 	// Waiting to pre-compute and persist latest summary would slow down document creation,
 	// so skip this step if it is a new document.
-	if (!isNew && isContainerSummary(payload)) {
-		const latestFullSummary: IWholeFlatSummary | undefined = await wholeSummaryManager
-			.readSummary(writeSummaryResponse.id)
-			.catch((error) => {
-				// This read is for Historian caching purposes, so it should be ignored on failure.
-				Lumberjack.error(
-					"Failed to read latest summary after writing container summary",
-					lumberjackProperties,
-					error,
-				);
-				return undefined;
-			});
-		if (latestFullSummary) {
-			if (persistLatestFullSummary) {
-				try {
-					// TODO: does this fail if file is open and still being written to from a previous request?
-					await persistLatestFullSummaryInStorage(
-						fileSystemManager,
-						getFullSummaryDirectory(
-							repoManager,
-							repoManagerParams.storageRoutingId.documentId,
-						),
-						latestFullSummary,
+	if (isContainerSummary(payload)) {
+		const latestFullSummary: IWholeFlatSummary | undefined = (
+			writeSummaryResponse as IWholeFlatSummary
+		).trees
+			? writeSummaryResponse
+			: await wholeSummaryManager.readSummary(writeSummaryResponse.id).catch((error) => {
+					// This read is for Historian caching purposes, so it should be ignored on failure.
+					Lumberjack.error(
+						"Failed to read latest summary after writing container summary",
 						lumberjackProperties,
+						error,
 					);
-				} catch (error) {
+					return undefined;
+			  });
+		if (latestFullSummary) {
+			if (persistLatestFullSummary) {
+				// Send latest full summary to storage for faster read access.
+				const persistP = persistLatestFullSummaryInStorage(
+					fileSystemManager,
+					getFullSummaryDirectory(
+						repoManager,
+						repoManagerParams.storageRoutingId.documentId,
+					),
+					latestFullSummary,
+					lumberjackProperties,
+				).catch((error) => {
+					// Persisting latest summary is an optimization, not a requirement, so do not throw on failure.
 					Lumberjack.error(
 						"Failed to persist latest full summary to storage during createSummary",
 						lumberjackProperties,
 						error,
 					);
-					// TODO: Find and add more information about this failure so that Scribe can retry as necessary.
-					throw new NetworkError(
-						500,
-						"Failed to persist latest full summary to storage during createSummary",
-					);
+				});
+				if (!isNew) {
+					// To avoid any possible race conditions when outside the critical path, we can await the persist operation.
+					// Chances for a race condition are slim and likely inconsequential, but better safe than sorry.
```

Comment:
> Is the race condition here that while this createSummary() is being completed, a getSummary() is called which would ask for this summary()? If yes, since we are returning the full summary to historian to cache, we don't have to worry about it right, because next requests will be served from cache?

Correct

> Also, as per my latest discussion with @yangg-msft , we were thinking that we could make this persisting of full summary a fire-n-forget for all cases,( whether its a new doc or not), mainly because we cache in Historian.

I don't really see what benefit that would have, but sure. We can discuss it for a future PR.